### PR TITLE
feat(dyn-filter): derive `condition_always_relax` from column monotonicity

### PR DESCRIPTION
**Pull request by stdrc**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Derive `condition_always_relax` property of DynamicFilter from column monotonicity property of plan node introduced in #17600.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/frontend/src/optimizer/plan_node/stream_exchange.rs [hunk 1]
```diff
@@ -34,17 +34,22 @@ pub struct StreamExchange {
 
 impl StreamExchange {
     pub fn new(input: PlanRef, dist: Distribution) -> Self {
-        // Dispatch executor won't change the append-only behavior of the stream.
+        let columns_monotonicity = if matches!(dist, Distribution::Broadcast) {
+            input.columns_monotonicity().clone()
+        } else {
+            // we lost monotonicity information when shuffling
+            MonotonicityMap::new()
+        };
```
Comment (xxchan) [inline src/frontend/src/optimizer/plan_node/stream_exchange.rs:42]: Does `Broadcast` requires the input to be singleton?

Comment (xxchan) [reply src/frontend/src/optimizer/plan_node/stream_exchange.rs:42]: Wait, so I guess actually it's when input distribution is `Singleton`, all exchange (not only Broadcast) don't lose monotonicity (if the shuffle _preserves order_).

Comment (stdrc) [reply src/frontend/src/optimizer/plan_node/stream_exchange.rs:42]: You're right.

**Review by xxchan — APPROVED** [review]

File: src/frontend/src/optimizer/plan_node/stream_project.rs [hunk 2]
```diff
@@ -82,19 +82,21 @@ impl StreamProject {
...
             use monotonicity_variants::*;
             match analyze_monotonicity(expr) {
-                Inherent(Constant) => {
-                    // XXX(rc): we can produce one watermark on each recovery for this case.
-                }
                 Inherent(monotonicity) => {
-                    if monotonicity.is_non_decreasing() {
+                    out_monotonicity_map.insert(expr_idx, monotonicity);
+                    if monotonicity.is_non_decreasing() && !monotonicity.is_constant() {
+                        // TODO(rc): may be we should also derive watermark for constant later
```
Comment (chenzl25) [inline src/frontend/src/optimizer/plan_node/stream_project.rs:92]: According to the usage pattern of `monotonicity.is_non_decreasing() && !monotonicity.is_constant()`, it seems using EnumAsInner is more straightforward.

**Review by chenzl25 — APPROVED** [review]
LGTM!

File: src/frontend/src/optimizer/plan_node/stream_project.rs (hunk 2 above)
Comment (stdrc) [reply src/frontend/src/optimizer/plan_node/stream_project.rs]: Actually we should have derive watermark for constant here as well, while for not changing irrelevant logic, I kept it the same as before.

Comment (chenzl25) [reply src/frontend/src/optimizer/plan_node/stream_project.rs]: IIUC, watermark should go up, but the constant watermark sounds trivial. Is there any use case in which we can benefit from a constant watermark?

Comment (stdrc) [reply src/frontend/src/optimizer/plan_node/stream_project.rs]: It's possible that we can benefit from getting a constant watermark for constant column each time we recover (if we don't derive watermark in this case, we won't know the watermark).

**Pull request merged**